File: packages/csharp/src/SheetCompressor/Encodings/InvertedIndex.cs
using System.Collections.Generic;
using System.Text;

namespace SheetCompressor.Encodings;

internal static class InvertedIndex
{
    /// <summary>
    /// Pack absolute (row, col) into a single long for HashSet/Dictionary keys.
    /// </summary>
    private static long Pack(int row, int col)
    {
        return ((long)row << 32) | (uint)col;
    }

    public static Encoding Encode(Grid grid, TokenCounter tokenCounter)
    {
        // Walk the grid in row-major order, bucketing every non-empty cell by
        // value. We use an ordered key list so iteration later yields values in
        // first-cell-address order (SPEC §4.4).
        var cellsByValue = new Dictionary<string, List<long>>();
        var valueOrder = new List<string>();

        for (int r = 0; r < grid.Rows.Count; r++)
        {
            var row = grid.Rows[r];
            for (int c = 0; c < row.Count; c++)
            {
                var value = row[c] ?? string.Empty;
                if (value.Length == 0) continue;
                var key = Pack(grid.Origin.Row + r, grid.Origin.Col + c);
                if (!cellsByValue.TryGetValue(value, out var bucket))
                {
                    bucket = new List<long>();
                    cellsByValue[value] = bucket;
                    valueOrder.Add(value);
                }
                bucket.Add(key);
            }
        }

        var groups = new List<InvertedIndexGroupJson>();

        foreach (var value in valueOrder)
        {
            var cellKeys = cellsByValue[value];
            var present = new HashSet<long>(cellKeys);
            var assigned = new HashSet<long>();
            var ranges = new List<string>();

            foreach (var startKey in cellKeys)
            {
                if (assigned.Contains(startKey)) continue;
                int startRow = (int)(startKey >> 32);
                int startCol = (int)(startKey & 0xFFFFFFFFu);

                // Maximum width: extend right while cells are in the value-set
                // AND not already absorbed by an earlier rectangle.
                int width = 1;
                while (true)
                {
                    var nk = Pack(startRow, startCol + width);
                    if (!present.Contains(nk) || assigned.Contains(nk)) break;
                    width++;
                }

                // Maximum height: extend down while every cell of the row of
                // `width` cells is still in the value-set and unassigned.
                int height = 1;
                while (true)
                {
                    int nextRow = startRow + height;
                    bool canExtend = true;
                    for (int dc = 0; dc < width; dc++)
                    {
                        var k = Pack(nextRow, startCol + dc);
                        if (!present.Contains(k) || assigned.Contains(k))
                        {
                            canExtend = false;
                            break;
                        }
                    }
                    if (!canExtend) break;
                    height++;
                }

                for (int dr = 0; dr < height; dr++)
                {
                    for (int dc = 0; dc < width; dc++)
                    {
                        assigned.Add(Pack(startRow + dr, startCol + dc));
                    }
                }

                var topLeft = Address.A1(startRow, startCol);
                if (width == 1 && height == 1)
                {
                    ranges.Add(topLeft);
                }
                else
                {
                    ranges.Add($"{topLeft}:{Address.A1(startRow + height - 1, startCol + width - 1)}");
                }
            }

            groups.Add(new InvertedIndexGroupJson { Value = value, Ranges = ranges });
        }

        var sb = new StringBuilder();
        for (int i = 0; i < groups.Count; i++)
        {
            if (i > 0) sb.Append('\n');
            var g = groups[i];
            for (int j = 0; j < g.Ranges.Count; j++)
            {
                if (j > 0) sb.Append('|');
                sb.Append(g.Ranges[j]);
            }
            sb.Append(',');
            sb.Append(Escape.EscapeValue(g.Value));
        }
        var s = sb.ToString();

        var json = new InvertedIndexJson
        {
            Origin = grid.Origin,
            Groups = groups,
        };
        return new Encoding { String = s, Json = json, TokenEstimate = tokenCounter(s) };
    }
}

File: packages/python/sheet_compressor/encodings/inverted_index.py
"""Inverted-index encoding (SPEC §4)."""

from typing import Any, Mapping

from ..address import a1
from .escape import escape_value


def encode_inverted_index(grid: Mapping[str, Any], token_counter) -> dict:
    rows = grid.get("rows", [])
    origin = grid["origin"]

    # Walk in row-major order, bucketing every non-empty cell by value.
    # `dict` preserves insertion order (Python 3.7+), so iterating later yields
    # values ordered by first cell address — exactly the order SPEC §4.4 wants.
    cells_by_value: "dict[str, list[tuple[int, int]]]" = {}
    for r in range(len(rows)):
        row = rows[r]
        for c in range(len(row)):
            value = row[c]
            if value == "":
                continue
            abs_row = origin["row"] + r
            abs_col = origin["col"] + c
            bucket = cells_by_value.get(value)
            if bucket is None:
                cells_by_value[value] = [(abs_row, abs_col)]
            else:
                bucket.append((abs_row, abs_col))

    groups = []
    for value, cell_coords in cells_by_value.items():
        present = set(cell_coords)
        assigned: set = set()
        ranges = []
        for start in cell_coords:
            if start in assigned:
                continue
            start_row, start_col = start

            # Maximum width: extend right while in the value-set and unassigned.
            width = 1
            while (
                (start_row, start_col + width) in present
                and (start_row, start_col + width) not in assigned
            ):
                width += 1

            # Maximum height: every cell of the next row of `width` cells must
            # still be in the value-set and unassigned.
            height = 1
            while True:
                next_row = start_row + height
                can_extend = True
                for dc in range(width):
                    k = (next_row, start_col + dc)
                    if k not in present or k in assigned:
                        can_extend = False
                        break
                if not can_extend:
                    break
                height += 1

            for dr in range(height):
                for dc in range(width):
                    assigned.add((start_row + dr, start_col + dc))

            top_left = a1(start_row, start_col)
            if width == 1 and height == 1:
                ranges.append(top_left)
            else:
                bottom_right = a1(start_row + height - 1, start_col + width - 1)
                ranges.append(f"{top_left}:{bottom_right}")

        groups.append({"value": value, "ranges": ranges})

    string = "\n".join(
        f"{'|'.join(g['ranges'])},{escape_value(g['value'])}" for g in groups
    )

    json_obj = {
        "encoding": "inverted-index",
        "version": 0,
        "origin": {"row": origin["row"], "col": origin["col"]},
        "groups": groups,
    }
    return {"string": string, "json": json_obj, "tokenEstimate": token_counter(string)}
